store.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright (c) 2024 Diagrid Inc.
+Licensed under the MIT License.
+*/
+
+package etcdcron
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	etcdclient "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/mirror"
+	"google.golang.org/protobuf/proto"
+)
+
+// The JobStore persists and reads jobs from Etcd.
+type JobStore interface {
+	Put(ctx context.Context, job Job) error
+	Delete(ctx context.Context, jobName string) error
+}
+
+type etcdStore struct {
+	etcdClient     *etcdclient.Client
+	kvStore        etcdclient.KV
+	partitioning   Partitioning
+	organizer      Organizer
+	putCallback    func(context.Context, Job) error
+	deleteCallback func(context.Context, string) error
+}
+
+type noStore struct {
+	putCallback    func(context.Context, Job) error
+	deleteCallback func(context.Context, string) error
+}
+
+func NoStore(
+	putCallback func(context.Context, Job) error,
+	deleteCallback func(context.Context, string) error) JobStore {
+	return &noStore{
+		putCallback:    putCallback,
+		deleteCallback: deleteCallback,
+	}
+}
+
+func (s *noStore) Put(ctx context.Context, job Job) error {
+	s.putCallback(ctx, job)
+	return nil
+}
+
+func (s *noStore) Delete(ctx context.Context, jobName string) error {
+	s.deleteCallback(ctx, jobName)
+	return nil
+}
+
+func NewEtcdJobStore(
+	ctx context.Context,
+	client *etcdclient.Client,
+	organizer Organizer,
+	partitioning Partitioning,
+	putCallback func(context.Context, Job) error,
+	deleteCallback func(context.Context, string) error) (JobStore, error) {
+	s := &etcdStore{
+		etcdClient:     client,
+		kvStore:        etcdclient.NewKV(client),
+		partitioning:   partitioning,
+		organizer:      organizer,
+		putCallback:    putCallback,
+		deleteCallback: deleteCallback,
+	}
+	err := s.init(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+func (s *etcdStore) init(ctx context.Context) error {
+	for _, partitionId := range s.partitioning.ListPartitions() {
+		// TODO(artursouza): parallelize this per partition.
+		partitionPrefix := s.organizer.JobsPath(partitionId) + "/"
+		partitionSyncer := mirror.NewSyncer(s.etcdClient, partitionPrefix, 0)
+		rc, errc := partitionSyncer.SyncBase(ctx)
+
+		for r := range rc {
+			for _, kv := range r.Kvs {
+				err := s.notifyPut(ctx, kv, s.putCallback)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		err := <-errc
+		if err != nil {
+			return err
+		}
+
+		s.sync(ctx, partitionPrefix, partitionSyncer)
```

Comment:
I had that at some point but the context can be cancelled. I can put it back but there is no point of using the cron without initializing the store, so the init() or Run() function just moves to another place. We don't want upper client to use a cron without having the store initialized - even if cron is not running. So, it would be too many things for the client of cron to be concerned about.

Reply:
@artursouza I hear you- where I'm coming from is that across Dapr we are using the `New (just sets up struct, maybe errors)`, `Run (blocks, long running, returns on context cancelled)`, `Close (interrupts Run and waits for Run and all resources to return/be released)`. This agreed paradigm and shared behaviour/entry point naming means it is easier to maintain the large number of long running routines across the project, and enable us to reuse the same tooling/utils to manage them all. A lot of that came out of the graceful shutdown work. Def not saying the implementation is necessarily wrong, but consistency is king here- least surprises reduces cognitive load to the reader (and person debugging!).
https://github.com/dapr/kit/blob/e33fbab745480b65aa54a512d6d743cc30525522/concurrency/closer.go#L67
https://github.com/dapr/dapr/blob/7c3011e7817dcc5b68ff9769964c2982b4102354/pkg/runtime/runtime.go#L258
https://github.com/dapr/dapr/blob/7c3011e7817dcc5b68ff9769964c2982b4102354/pkg/runtime/hotreload/loader/disk/disk.go#L71
https://github.com/dapr/dapr/blob/7c3011e7817dcc5b68ff9769964c2982b4102354/pkg/sentry/sentry.go#L121
https://github.com/dapr/dapr/blob/7c3011e7817dcc5b68ff9769964c2982b4102354/pkg/runtime/hotreload/hotreload.go#L90

There is an argument to be made about reducing diffs especially in behavior, as this is a forked library, though I think there is an assumed agreement we are going to be highly deviating from source and will need to maintain this long term anyway.

Reply:
OK. We are deviating from the original design anyway. I created an Init() method now.